roadmap/implementers-guide/src/node/utility/availability-store.md

```diff
@@ -39,130 +33,137 @@ digraph {
+  Finalized(Timestamp)
+}
+```
 
-On `OverseerSignal::BlockFinalized(_)` events:
+We maintain the invariant that if a candidate has a meta entry, its available data exists on disk if `data_available` is true. All chunks mentioned in the meta entry are available.
 
-- Handle all pruning based on the newly-finalized block.
+Additionally, there is exactly one `prune_by_time` entry which holds the candidate hash unless the state if `Unfinalized`. There may be zero, one, or many "unfinalized" keys with the given candidate, and this will correspond to the `state` of the meta entry.
 
-On `QueryPoV` message:
+## Protocol
 
-- Return the PoV block, if any, for that candidate hash.
+Input: [`AvailabilityStoreMessage`][ASM]
 
-On `QueryChunk` message:
+Output:
+- [`RuntimeApiMessage`][RAM]
 
-- Determine if we have the chunk indicated by the parameters and return it and its inclusion proof via the response channel if so.
 
-On `StoreChunk` message:
+## Functionality
 
-- Store the chunk along with its inclusion proof under the candidate hash and validator index.
+For each head in the `activated` list:
+  - Note any new candidates backed in the block. Update the `CandidateMeta` for each. If the `CandidateMeta` does not exist, create it as `Unavailable` with the current timestamp. Register a `"prune_by_time"` entry based on the current timestamp + 1 hour.
+  - Note any new candidate included in the block. Update the `CandidateMeta` for each, performing a transition from `Unavailable` to `Unfinalized` if necessary. That includes removing the `"prune_by_time"` entry. Add the block hash and number to the state, if unfinalized. Add an `"unfinalized"` entry for the block and candidate.
+  - The `CandidateEvent` runtime API can be used for this purpose.
+  - TODO: load all ancestors of the head back to the finalized block so we don't miss anything if import notifications are missed. If a `StoreChunk` message is received for a candidate which has no entry, then we will prematurely lose the data.
 
-On `StorePoV` message:
+On `OverseerSignal::BlockFinalized(finalized)` events:
+  - for each key in `iter_by_prefix("unfinalized")`
+    - Stop if the key is beyond `("unfinalized, finalized)`
+    - For each block number f that we encounter, load the finalized hash for that block.
+      - For each candidate that we encounter under `f` and the finalized block hash,
+        - Update the `CandidateMeta` to have `State::Finalized`. Remove the existing `"prune_by_time"` entry. If the previous state was `Unfinalized`, remove all `"unfinalized"` entries.
+        - Register a `"prune_by_time"` entry for the candidate based on the current time + 1 day + 1 hour.
+      - For each candidate that we encounter under `f` which is not under the finalized block hash,
+        - Remove all entries under `f` in the `CandidateMeta`.
+        - If the `CandidateMeta` has state `Unfinalized` with an empty list of blocks, downgrade to `Unavailable` and 
```

Comment:
It is pruned because when we transition from `Unavailable -> Unfinalized`, we keep track of the same timestamp where we first stored the data. When we proceed from `Unfinalized -> Unavailable` we re-register the pruning under that same timestamp. 

I don't think we want to prune it directly because if the candidate appears in block `A1` but then `A2` is finalized, we still want the candidate to be able to appear in `B2`. So we don't want to greedily prune after A2 is finalized. This can not happen right now but could happen with contextual execution, where the candidate has relay-parent `G`. That is why I just downgrade to the `Unavailable` state for  clean-up to happen later.

```rust
G -> A1
G -> A2 -> B2
```

Reply:
I realize that there is a trailing sentence here and have updated it.